lib/gc/Transforms/TilingUsingInterfaceX.cpp

```diff
@@ -0,0 +1,987 @@
+//===-- TilingUsingInterfaceX.cpp -  upstream eXtension ---------*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/SCF/Transforms/TileUsingInterface.h"
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Affine/Utils.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Arith/Utils/Utils.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Linalg/Utils/Utils.h"
+#include "mlir/Dialect/SCF/Utils/Utils.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Utils/IndexingUtils.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Interfaces/DestinationStyleOpInterface.h"
+#include "mlir/Interfaces/TilingInterface.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/Debug.h"
+#include <optional>
+
+#include "TilingUsingInterfaceX.h"
+
+#define DEBUG_TYPE "tile-using-interface-x"
+
+using namespace mlir;
+
+static Operation *cloneOpAndUpdateDestinationArgs(RewriterBase &rewriter,
+                                                  Operation *op,
+                                                  ValueRange newDestArgs) {
+  Operation *clonedOp = rewriter.clone(*op);
+  if (newDestArgs.empty())
+    return clonedOp;
+  if (auto destinationStyleOp = dyn_cast<DestinationStyleOpInterface>(clonedOp))
+    destinationStyleOp.getDpsInitsMutable().assign(newDestArgs);
+  return clonedOp;
+}
+
+static std::tuple<OpResult, std::optional<OpOperand *>>
+getUntiledProducerFromSliceSource(OpOperand *source,
+                                  ArrayRef<LoopLikeOpInterface> loops) {
+  std::optional<OpOperand *> destinationIterArg;
+  auto loopIt = loops.rbegin();
+  while (auto iterArg = dyn_cast<BlockArgument>(source->get())) {
+    auto loop = *loopIt;
+    if (iterArg.getOwner()->getParentOp() != loop)
+      break;
+    source = loop.getTiedLoopInit(iterArg);
+    loopIt++;
+  }
+  if (loopIt == loops.rend())
+    destinationIterArg = source;
```

Comment:
First of all, this is also current upstream [status](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp#L1066).  Since parts of necessary utility are marked `static` in upstream code, we have to copy them downstream for further extension. And that is why I suggest current review might better bypass these change from this file. It is welcomed to go through corresponding upstream PR to see the real changes compared with MLIR master(See https://github.com/llvm/llvm-project/pull/94190 and https://github.com/llvm/llvm-project/pull/97803).

Anyway, I agree with you here, So lets add emptiness check downstream first.

Reply:
Not a blocker, we can clean it up later together with the upstream changes.